source-mongodb/pull.go

```diff
@@ -412,25 +482,25 @@ func (c *capture) BackfillCollection(ctx context.Context, client *mongo.Client,
 			if err = c.Output.Checkpoint(checkpointJson, true); err != nil {
 				return fmt.Errorf("output checkpoint failed: %w", err)
 			}
+		}
 
-			if diff, err := oplogTimeDifference(ctx, client); err != nil {
-				return fmt.Errorf("could not read oplog, access to oplog is necessary: %w", err)
-			} else {
-				// TODO: This is just spamming logs, but users won't see these logs. Should we
-				// just error after seeing this error for a while?
-				if diff < minOplogTimediff {
-					log.Warn(fmt.Sprintf("the current time difference between oldest and newest records in your oplog is %d seconds. This is smaller than the minimum of 24 hours. Please resize your oplog to be able to safely capture data from your database: https://go.estuary.dev/NurkrE", diff))
-				}
+		// This means the next call to cursor.Next() will trigger a new network
+		// request. Before we do that, we release our acquired semaphore and try to
+		// acquire a new one, to give chance to other bindings to progress in their
+		// backfill
+		if cursor.RemainingBatchLength() == 0 {
+			sp.Release(1)
+			if err := sp.Acquire(ctx, 1); err != nil {
+				return fmt.Errorf("semaphore acquisition ctx cancelled: %w", err)
 			}
 		}
 	}
 
-
 	if err := cursor.Err(); err != nil {
 		return fmt.Errorf("cursor error for backfill %s: %w", res.Collection, err)
 	}
 
-	state.Status = StatusStreaming
+	state.Backfill.Done = true
 	var checkpoint = captureState{
 		Resources: map[string]resourceState{
 			resourceId(res): *state,
```

Comment:
Nit: This function could take a `resourceState` directly (not a reference), to avoid needing to do this dereference.